### DISCUSSION TITLE
[2.8.0] - load_balancer name is already used

### DISCUSSION THREAD

**Discussion by K-MeLeOn** [mysticaltech/terraform-hcloud-kube-hetzner#1028] [Q&A] [answered]
Hi everyone !

After upgrade to the last version (2.8.0), I update my kube.tf file than just some commentary was added, I use `terraform apply --auto-approve` command and got this error :

> module.kube-hetzner.hcloud_load_balancer.cluster[0]: Creating...
╷
│ Error: name is already used (uniqueness_error)
│ 
│   with module.kube-hetzner.hcloud_load_balancer.cluster[0],
│   on .terraform/modules/kube-hetzner/init.tf line 1, in resource "hcloud_load_balancer" "cluster":
│    1: resource "hcloud_load_balancer" "cluster" {
│ 
╵

No config change in the kube.tf file, I don't know where to search, any idea ?

Thanks for you time !!

**Comment by janis-ax** [comment]
I have the same issue! 👍 

```bash
Terraform used the selected providers to generate the following execution plan. Resource actions are indicated with the following symbols:
  + create

Terraform will perform the following actions:

  # module.kube-hetzner.hcloud_load_balancer.cluster[0] will be created
  + resource "hcloud_load_balancer" "cluster" {
      + delete_protection  = false
      + id                 = (known after apply)
      + ipv4               = (known after apply)
      + ipv6               = (known after apply)
      + labels             = {
          + "cluster"     = "my-k3s"
          + "engine"      = "k3s"
          + "provisioner" = "terraform"
        }
      + load_balancer_type = "lb11"
      + location           = "fsn1"
      + name               = "my-k3s"
      + network_id         = (known after apply)
      + network_ip         = (known after apply)
      + network_zone       = (known after apply)

      + algorithm {
          + type = "round_robin"
        }
    }

  # module.kube-hetzner.null_resource.kustomization will be created
  + resource "null_resource" "kustomization" {
      + id       = (known after apply)
      + triggers = {
          + "helm_values_yaml" = (sensitive value)
          + "options"          = ""
          + "versions"         = <<-EOT
                v1.27
                v1.27.3
                N/A
                N/A
                N/A
                N/A
                v1.14.0
            EOT
        }
    }

Plan: 2 to add, 0 to change, 0 to destroy.

───────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────

Saved the plan to: tf.plan

To perform exactly these actions, run the following command to apply:
    terraform apply "tf.plan"
 terraform apply "tf.plan"  
module.kube-hetzner.hcloud_load_balancer.cluster[0]: Creating...
╷
│ Error: name is already used (uniqueness_error)
│ 
│   with module.kube-hetzner.hcloud_load_balancer.cluster[0],
│   on .terraform/modules/kube-hetzner/init.tf line 1, in resource "hcloud_load_balancer" "cluster":
│    1: resource "hcloud_load_balancer" "cluster" {
```

**Reply by Silvest89** [reply]
@mysticaltech 
https://github.com/kube-hetzner/terraform-hcloud-kube-hetzner/pull/1001

This is the root issue. While the author did test creating everything from scratch. He didn't account for existing clusters which already have a load balancer with that same name. (HCloud managed)

To fix it either:
* Suffix/prefix to the pre-created loadbalancer name (also change it in annotations code piece
* Add an enabled attribute for backwards compatibility (eg TF managed, or HCloud managed)

Also:
- If Terraform managed deletion protection should be enabled so Hcloud cannot accidently delete it

Anyways @K-MeLeOn  @janis-ax you can delete the load balancer in the Hetzner Console and then run terraform apply this should make the loadbalancer terraform managed (it wasn't previously)

**Reply by mysticaltech** [reply]
Exactly @Silvest89, thanks for pointing this out.  @schlichtanders Do you have the bandwith to publish a fix based on the above recommendations? It would be awesome 🙏

**Reply by mysticaltech** [reply]
@Silvest89  @schlichtanders I implemented the fix, at least the start of it. https://github.com/kube-hetzner/terraform-hcloud-kube-hetzner/pull/1034. A review would be appreciated folks . 

@Silvest89 I followed your proposed fix. It should let the upgrade go through, but I am not sure that it will replace the old one, probably not, and probably it's a good thing to let people migrate slowly as they see fit. But your opinion would be very much appreciated!

**Reply by Silvest89** [reply]
@mysticaltech 
Well the hcloud controller should actually delete the old one if it isn't referenced anymore. Since you have a new name now. This should be a backwards compatible fix. There is one thing missing though I will comment it on the PR

**Reply by schlichtanders** [reply]
@Silvest89  and @mysticaltech thank you so much for fixing the pullrequest of mine :pray: .

**Comment by mysticaltech** [comment, marked as answer]
@K-MeLeOn @janis-ax FYI folks, this issue is now fixed in v2.8.2. Thanks again for reporting it.

**Reply by dstockhammer** [reply]
Hi @mysticaltech, just to confirm; will v2.8.3 now cleanly/silently upgrade from v2.7 or are the manual steps still required? 
Thanks!

**Reply by mysticaltech** [reply]
@dstockhammer Pertinent question. Manual steps are required, basically what was done above by @K-MeLeOn.

1. Delete the old LB.
2. Get the new IP with kubectl get svc -A.
3. Point your DNS to the new IP.

**Reply by dstockhammer** [reply]
Hi @mysticaltech, I finally got to upgrade the cluster. Since Hetzner has an outage/service degradation anyway, I figured now is a good opportunity to potentially take the LB out 🙃 

I'm a bit confused at the behaviour I'm seeing: A new LB was created next to the old/current one as `myname-traefik`, and at the end of the `terraform apply`, the IPv4 and IPv6 addresses of the **new** LB are reported (`ingress_public_ipv4` and `ingress_public_ipv6`). I think this is expected. However, the new LB does not have any service targets and no internal IP. The old LB still has targets. DNS is still pointing to the old LB and when I switch over to the new LB, it doesn't work (as expected, since it has no targets). Using the old LB still works.

Did something go wrong during the update? 🤔

Maybe it's worth nothing that I'm using `lb_hostname` and `kubectl get svc -A` reports that hostname as "external IP". I tried removing `lb_hostname`, since it's not _really_ required, but then the init times out while `Waiting for load-balancer to get an IP...`.

I get the following warning from the `traefik` service:
```
  Type     Reason                  Age                 From                Message
  ----     ------                  ----                ----                -------
  Normal   EnsuringLoadBalancer    58s (x35 over 10d)  service-controller  Ensuring load balancer
  Warning  SyncLoadBalancerFailed  57s (x15 over 36m)  service-controller  Error syncing load balancer: failed to ensure load balancer: hcloud/loadBalancers.EnsureLoadBalancer: hcops/LoadBalancerOps.ReconcileHCLB: hcops/LoadBalancerOps.changeHCLBInfo: name is already used (uniqueness_error)
```

Does that mean I have to delete the old LB **before** the new one can get the targets assigned etc? I'm a bit hesitant to do that right now since there's no coming back from that :S

Thanks!

**Reply by mysticaltech** [reply]
@dstockhammer Yes, you got it right, you have to delete the old LB point all dns to it and it should all fall into place! ;)

**Reply by dstockhammer** [reply]
Sorry I'm not ready to do that to the production cluster, though I do realise that by applying this I may already have passed the point of no return :(

What difference does it make to "the infrastructure" that a different LB already exists? I don't understand that. If it wasn't able to attach the new LB to to the network and assign targets? What I tried instead was the following:
- Deleted `myname-traefik` (that's the new LB that was created just now)
- Detached `myname` (that's the old/existing LB) from the network. **At this point the service is down**.
- Applied Terraform once more; the script recreated `myname-traefik` same as above; and again didn't attach it to the network and assigned no services. Why? Does existence of the `myname` LB really make a difference? It's completely unattached to anything Kube-related at this point 🤔 



PS - Thanks for the quick reply, I really appreciate it!

**Reply by dstockhammer** [reply]
Update: OK I don't know if this is going to lead to any problems down the road, but I decided to [import](https://registry.terraform.io/providers/hetznercloud/hcloud/latest/docs/resources/load_balancer#import) the existing LB instead of creating a new one and having to delete the old one. 

```
> terragrunt state rm 'module.kube-hetzner.hcloud_load_balancer.cluster[0]'
> terragrunt import 'module.kube-hetzner.hcloud_load_balancer.cluster[0]' 0000
> terragrunt apply

Terraform will perform the following actions:

  # module.kube-hetzner.hcloud_load_balancer.cluster[0] will be updated in-place
  ~ resource "hcloud_load_balancer" "cluster" {
        id                 = "0000000"
      ~ labels             = {
          + "cluster"                = "xxx"
          + "engine"                 = "k3s"
          - "hcloud-ccm/service-uid" = "redacted" -> null
          + "provisioner"            = "terraform"
        }
      ~ name               = "xxx" -> "xxx-traefik"
        # (8 unchanged attributes hidden)

        # (4 unchanged blocks hidden)
    }
```

So far it looks like the old LB has been picked up successfully and everything **_would_** have worked without downtime. If everything does indeed continue working, that is...

**Reply by mysticaltech** [reply]
@dstockhammer  Well done, happy you solved it that way. No reason why it should not work. IMHO.

**Reply by dstockhammer** [reply]
One more update; so far it looks like everything is working, except for one thing: New nodes aren't added as targets in the load balancer, regardless of whether they're provisioned manually via a static node pool or by the autoscaler. I solved that issue by **manually** editing the load balancer to select targets by label. The labels are different for static and autoscaled nodes, but `role=agent_node` and `hcloud/node-group=xxx` (the value of this label depends on the configured name of the node group) selects all the workers.

That seems to work fine for me. I'll let you know if I run into any issues in case anyone wants to copy this approach.

**Reply by mysticaltech** [reply]
@dstockhammer  So it's not playing nice with the Hetzner cloud controller.

**Comment by K-MeLeOn** [comment]
Nice !
Thank you @Silvest89 and @mysticaltech.
After removing the LB (and change the IP address of the LB domain to the newly created IP address), everything works. :+1:

[Version 2.8.2]

**Comment by Lennix** [comment]
So I don't know what happened here, but it's still a mess.

I've tried to upgrade multiple times now and I always get stuck at the load balancer. Terraform creates the new load-balancer but it stays disconnected.

I finally could get my old setup working again. Somehow hcloud-cloud-controller-manager had `HCLOUD_LOAD_BALANCERS_ENABLED` set to false. Was that triggered by the update? Or might that be the cause of the new load-balancer not connecting?

**Reply by K-MeLeOn** [reply]
Did you remove the LB before ugrading ?

**Reply by Lennix** [reply]
I removed the LBs countless times. They also were named differently.

**Reply by K-MeLeOn** [reply]
Strange behavior, we probably need your kube.tf file (remove sensitive data) to understand.
You don't have a terraform error, right ?

**Reply by Lennix** [reply]
Terraform regularly timed out.

I am quite certain it was caused by `HCLOUD_LOAD_BALANCERS_ENABLED` - Now that I've enabled it and retried a couple times the lb is now correctly created and connected to the nginx ingress

**Reply by wolskikd** [reply]
@Lennix how did you enabled that? By setting `enable_klipper_metal_lb` to false or in a manual way? I have similar issues during upgrade of kube-hetzner, new LB has new name and no targets at all

**Reply by Lennix** [reply]
I set it manually by adjusting the deployment manifest (using Openlens)
